File: python/semanticabi/steps/TransformErrorStep.py
from typing import List, Dict

from semanticabi.metadata.EthBlock import EthBlock
from semanticabi.metadata.EthTransaction import EthTransaction
from semanticabi.steps.AbiSchema import AbiSchema
from semanticabi.steps.Step import Step, TransformItem, TRANSFORM_ERROR_COLUMN
from semanticabi.steps.SubsequentStep import SubsequentStep


class TransformErrorStep(SubsequentStep):
    """
    Special step that adds a column to the schema for holding errors that occurred during the transform. This doesn't
    actually add the error to the data, but only informs the final transform process that it should include any errors
    in the data
    """

    _schema: AbiSchema

    def __init__(self, previous_step: Step):
        super().__init__(previous_step)
        self._schema = previous_step.schema.with_columns([TRANSFORM_ERROR_COLUMN])

    @property
    def schema(self) -> AbiSchema:
        return self._schema

    def _inner_transform_item(
        self,
        block: EthBlock,
        transaction: EthTransaction,
        item: TransformItem,
        previous_data: List[Dict[str, any]]
    ) -> List[Dict[str, any]]:
        return previous_data

File: js/src/semanticabi/transform/steps/TransformErrorStep.ts
import {SubsequentStep} from "semanticabi/transform/steps/SubsequentStep";
import {Step} from "semanticabi/transform/steps/Step";
import {AbiSchema} from "semanticabi/transform/steps/AbiSchema";
import {EthBlock} from "semanticabi/metadata/EthBlock";
import {EthTransaction} from "semanticabi/metadata/EthTransaction";
import {TransformItem} from "semanticabi/transform/steps/TransformItem";
import {JsonObject} from "common/CommonTypes";

/**
 * Special step that adds a column to the schema for holding errors that occurred during the transform. This doesn't
 * actually add the error to the data, but only informs the final transform process that it should include any errors
 * in the data.
 */
export class TransformErrorStep extends SubsequentStep {

    public readonly schema: AbiSchema;

    constructor(previousStep: Step) {
        super(previousStep);

        this.schema = this.previousStep.schema.withColumns([Step.TRANSFORM_ERROR_COLUMN]);
    }

    protected innerTransformItem(
        block: EthBlock,
        transaction: EthTransaction,
        item: TransformItem,
        previousData: JsonObject[]
    ): JsonObject[] {
        return previousData;
    }

}